altapay.php

```diff
@@ -0,0 +1,2390 @@
+                        ],
+                    ],
+                ],
+            ],
+            'submit'  => [
+                'title' => $this->l('Save'),
+            ],
+            'buttons' => [
+                [
+                    'href'  => AdminController::$currentIndex .
+                               '&configure=' . $this->name . '&token=' . Tools::getAdminTokenLite('AdminModules'),
+                    'title' => $this->l('Back to list'),
+                    'icon'  => 'process-icon-back'
+                ]
+            ],
+        ];
+        $helper                        = new HelperForm();
+        $helper->module                = $this;
+        $helper->name_controller       = 'altapay';
+        $helper->identifier            = $this->identifier;
+        $helper->token                 = Tools::getAdminTokenLite('AdminModules');
+        $helper->currentIndex          = AdminController::$currentIndex . '&configure=' . $this->name;
+        $helper->show_toolbar          = false;
+        $helper->table                 = 'altapay_terminals';
+        $lang                          = new Language((int)Configuration::get('PS_LANG_DEFAULT'));
+        $helper->default_form_language = $lang->id;
+        $helper->id                    = (int)Tools::getValue('id_terminal');
+        $helper->submit_action         = 'savealtapay_terminals';
+        $helper->tpl_vars              = [
+            'fields_value' => (array)$this->getFormValues(),
+            'languages'    => (array)$this->context->controller->getLanguages(),
+            'id_language'  => (array)$this->context->language->id
+        ];
+
+        return $helper->generateForm($fieldsForm);
+    }
+
+    /**
+     * Query the AltaPay API for available terminals
+     *
+     * @param bool $objects
+     *
+     * @return array<int, Terminal>
+     */
+    private function getAltapayTerminals($objects = false)
+    {
+        require_once(_PS_MODULE_DIR_ . '/altapay/lib/AltapayMerchantAPI.class.php');
+        $cgConf                = [];
+        $terminalArray         = [];
+        $termNature            = '';
+        $cgConf['user']        = $this->getAPIUsername();
+        $cgConf['password']    = $this->getAPIPassword();
+        $cgConf['altapay_url'] = $this->getAltapayUrl();
+        $api                   = null;
+        try {
+            $api               = new AltapayMerchantAPI($cgConf['altapay_url'], $cgConf['user'], $cgConf['password'],
+                null);
+            $response          = $api->login();
+            $responseTerminals = $api->getTerminals();
+            $terminals         = $responseTerminals->getTerminals();
+            if (!$response->wasSuccessful()) {
+                $resErrMsg  = $response->getErrorMessage();
+                $resErrCode = $response->getErrorCode();
+                throw new Exception('Could not login to the Merchant API: ' . $resErrMsg, $resErrCode);
```

Comment:
Could we use an exception type that is specific to our module so that it is easier to track where the issue is coming from?

Reply:
This has been updated. The exception contains the plugin name for tracking the issue.

Reply:
You changed the message, but not the exception type. What I'm suggestion:
```php
throw new AltapayException('Could not login to the Merchant API: ' . $resErrMsg, $resErrCode);
```